attest/activation.go

```diff
@@ -88,32 +98,68 @@ func (p *ActivationParameters) checkTPM12AKParameters() error {
 	return nil
 }
 
+// VerifyOpts specifies options passed to (*AttestationParameters).Verify()
+type VerifyOpts struct {
+	// SelfAttested set to true ensures that the attestation is self-signed,
+	// set to false ensures that the attestation is not self-signed.
+	SelfAttested bool
+	// Restricted set to true ensures that the verified key was created by TPM
+	// with the restricted flag; set to false ensures that the flag was not set.
+	Restricted bool
+}
+
 func (p *ActivationParameters) checkTPM20AKParameters() error {
-	if len(p.AK.CreateSignature) < 8 {
-		return fmt.Errorf("signature is too short to be valid: only %d bytes", len(p.AK.CreateSignature))
+	// AK must be restricted and its attestation is self-signed
+	opts := VerifyOpts{
+		SelfAttested: true,
+		Restricted:   true,
+	}
+	return p.AK.Verify(opts)
+}
+
+// Verify verifies the TPM2-produced attestation parameters checking whether:
+// - the attestation is self-signed (specified by opts)
+// - the key length is secure
+// - the attestation parameters matched the attested key
+// - the key was TPM-generated and resides within TPM
+// - the key can or cannot sing outside TPM-objects (specified via opts)
+// - the signature is correct
+func (p *AttestationParameters) Verify(opts VerifyOpts) error {
```

Comment:
The way this is laid out, it's really hard to see what's being changed for "Generate()" or what data is verified by the signature in "Verify()". Also if any of this is wrong, that's an auth bypass :)

I also don't see the reason we want to check the signature for Generate(), that seems redundant?

I would have expected logic like

- Generate()
  - check key attributes
  - generate encrypted challenge
- Verify()
  - check signature
  - check key attributes

Where the "check key attributes" part is its own function separately called by Generate() and Verify(). That makes it clear what is common, but ignore other components like "SelfAttested"?

Reply:
Realize this may be a historical issue rather than an issue with this PR. @twitchy-jsonp if he has any thoughts on how we might make this a little cleaner and easier to verify it's correct.

Reply:
I'm not sure are we on the same page. The Generate()/Verify() logic is for the credential activation protocol, involving ActivationParameters (with a challenge, AttestationParameters,...). AttestationParameters' verification is only part of that (as mentioned, AttestationParameter's Verify for this case must set both options SelfSigned and Restricted to true)

Key certification is non-interactive: client/prover generates `ap=sk.AttestationParameters()` and server/verifier just calls `ap.Verify(...)` (SelfSigner=Restricted=false) and makes sure `ap.CertifyingKey` is trusted.